File: Functions/SaveToStorage.js
//This function connects to the MySQL Instance and does a dump per character to Bucker
exports.buildFiles = (req, res) => {
	// Parameters
	var db = 'bestbuy';
	var pwd = 'Set2=$u^qV4#5);*';
	var user = 'data_reader';
	var dbhost = '35.224.221.233';
	var targetBucket = 'rgreaves-mysite';


	//Build Connection
	var mysql = require('mysql');
	var con = mysql.createConnection({
	  host: dbhost,
	  user: user,
	  password: pwd,
	  database: db
	});

	var BreakDown = [ 'A','B','C','D','E','F','G','H','I','J','K','L','M','N','O','P','Q','R','S','T','U','V','W','X','Y','Z' ];

	// Itterate connection to build file
	con.connect(function(err) {
	  if (err) throw err;
	  console.log("Connected!");
	  //Itterate through array
	  BreakDown.forEach(function(value){
		var sql = 	"SELECT 'name','price','category_name','shipping','description','manufacturer','model','url','image' FROM products WHERE 'name' like '" + value + "%'";
		con.query(sql, function (err, result) {
		if (err) throw err;
			console.log(result);
	  });
	  }

	});

};

function SaveToBucket(WhichArray,objectName,bucketName) {
	
	var storage = require('@google-cloud/storage')();
	var myBucket = storage.bucket(bucketName);

	var options = {
	  entity: 'allUsers',
	  role: storage.acl.READER_ROLE
	};

	myBucket.acl.add(options, function(err, aclObject) {});

	myBucket.acl.add(options).then(function(data) {
	  var aclObject = data[0];
	  var apiResponse = data[1];
	});
}

File: Functions/SaveToStorage.py
#!/usr/bin/python

### Import Modules
import os
import datetime
import json
import mysql.connector
import string
import subprocess
from subprocess import call

targetBucket = 'rgreaves-mysite';
hostname = '35.224.221.233'
username = 'data_reader'
password = 'Set2=$u^qV4#5);*'
database = 'bestbuy'

# Query the DB for each character
def doQuery(connection,searchChar=0) :
    cur = connection.cursor(buffered=True)

    #Run Query
    if searchChar != 0:
       cur.execute("SELECT distinct(name) FROM products WHERE name like '" + searchChar + "%'")
    else:
       cur.execute( "SELECT distinct(name) FROM products WHERE name REGEXP '^[0-9]+'" )
       searchChar = "1"
    #Build Output array
    items = []
    for row in cur:
        items.append({'name':row[0]})
	
    #Only write log if value > 0
    if cur.rowcount > 0 :
        print ("Found " + str(cur.rowcount) + " records for " + searchChar + ", moving to next step")
        CreateJson( items, searchChar )
    else :
        print ("Found " + str(cur.rowcount) + " records for " + searchChar + ", moving to next char")

    cur.close()

# Query To Write to Temp local file
def CreateJson( ArrayOfData, FileName ) :
    FileName+= ".json"
    with open(FileName, 'w') as outfile:
      json.dump(ArrayOfData, outfile)
    
    #Upload file
    upload_files(FileName)
    #Remove file
    print("Removing Old File " + FileName)
    os.remove(FileName)
    
    return FileName

# Upload to Google Cloud Store
def upload_files(fileName):
    print("Uploading file " + fileName + " to gs://"+ targetBucket + "/pre-compiled/")
    p = subprocess.Popen('gsutil -h "Content-Type:application/json" cp '+ fileName +' gs://'+ targetBucket + '/pre-compiled/', shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
    retval = p.wait()

# Run Against Alaphabet
print(str(datetime.datetime.now()) + " -- Running batch task to pre-compile to first letter based JSON files")
myConnection = mysql.connector.connect( host=hostname, user=username, passwd=password, db=database )
for val in string.ascii_lowercase:
    doQuery( myConnection, val )
#Once more for none Alphabet characters
doQuery( myConnection )

#Kill Connection
myConnection.close()

